tests/integ/modin/frame/test_cache_result.py

```diff
@@ -0,0 +1,159 @@
+    with SqlCounter(query_count=1):
+        assert_empty_snowpark_pandas_equals_to_pandas(cached_snow_df, native_df)
+
+    native_df = native_pd.DataFrame(columns=["A", "B", "C"])
+    snow_df = pd.DataFrame(native_df)
+    with SqlCounter(query_count=1):
+        cached_snow_df = cache_and_return_df(snow_df, inplace)
+    if not inplace:
+        # If inplace, then cached_snow_df IS snow_df, so this check
+        # is meaningless.
+        with SqlCounter(query_count=1):
+            assert np.all((cached_snow_df == snow_df).values)
+    with SqlCounter(query_count=1):
+        assert_empty_snowpark_pandas_equals_to_pandas(cached_snow_df, native_df)
+
+    native_df = native_pd.DataFrame(index=["A", "B", "C"])
+    snow_df = pd.DataFrame(native_df)
+    with SqlCounter(query_count=1):
+        cached_snow_df = cache_and_return_df(snow_df, inplace)
+    if not inplace:
+        # If inplace, then cached_snow_df IS snow_df, so this check
+        # is meaningless.
+        with SqlCounter(query_count=1):
+            assert np.all((cached_snow_df == snow_df).values)
+    with SqlCounter(query_count=1):
+        assert_empty_snowpark_pandas_equals_to_pandas(cached_snow_df, native_df)
+
+    native_df = native_pd.DataFrame(columns=["A", "B", "C"], index=[0, 1, 2])
+    snow_df = pd.DataFrame(native_df)
+    with SqlCounter(query_count=1):
+        cached_snow_df = cache_and_return_df(snow_df, inplace)
+    if not inplace:
+        # If inplace, then cached_snow_df IS snow_df, so this check
+        # is meaningless.
+        with SqlCounter(query_count=1):
+            assert np.all((cached_snow_df.isnull() == snow_df.isnull()).values)
+    with SqlCounter(query_count=1):
+        assert_snowpark_pandas_equals_to_pandas_without_dtypecheck(
+            cached_snow_df, native_df
+        )
+
+
+@pytest.mark.parametrize("inplace", [True, False])
+def test_cache_result_dataframe_complex_correctness(
+    time_index_string_column_snowpark_pandas_df,
+    time_index_string_column_native_df,
+    inplace,
+):
+    snow_df = time_index_string_column_snowpark_pandas_df
+    native_df = time_index_string_column_native_df
+
+    snow_df = snow_df.resample("2H").mean()
+    with SqlCounter(query_count=1):
+        cached_snow_df = cache_and_return_df(snow_df, inplace)
+    if not inplace:
+        # If inplace, then cached_snow_df IS snow_df, so this check
+        # is meaningless.
+        with SqlCounter(query_count=1):
+            assert np.all((cached_snow_df == snow_df).values)
+    native_df = native_df.resample("2H").mean()
+
+    cached_snow_df = cached_snow_df.set_index("b", drop=False)
+    native_df = native_df.set_index("b", drop=False)
+    with SqlCounter(query_count=1):
+        assert_snowpark_pandas_equals_to_pandas_without_dtypecheck(
+            cached_snow_df, native_df, check_freq=False
+        )
+
+
+@pytest.mark.parametrize("inplace", [True, False])
+class TestCacheResultReducesLatency:
```

Comment:
It seems too brittle to hope that we will consistently reduce latency. I think overall query count and count of things like `join`, `udtf`, etc. is more helpful. I suggest that we have a few test cases for different scenarios where users might want to cache result and that we show how caching reduces query counts compared to not caching. See @sfc-gh-lspiegelberg 's comment [here](https://github.com/snowflakedb/snowpark-python/pull/1536#discussion_r1594326009).